pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java

```diff
@@ -254,7 +254,11 @@ public String getTaskType() {
         // but the metrics are not available until the controller schedules a valid task
         long maxEndTimeMs = Long.MIN_VALUE;
         for (SegmentZKMetadata preSelectedSegment : preSelectedSegments) {
-          maxEndTimeMs = Math.max(maxEndTimeMs, preSelectedSegment.getEndTimeMs());
+          long currentEndTimeMs = preSelectedSegment.getEndTimeMs();
+          // Compute maxEndTimeMs among segments that are valid for merge (e.g. endTimeStamp < currentTime - bufferTime)
+          if (isValidBucketEndTime(currentEndTimeMs, bufferMs, lowerMergeLevel, mergeRollupTaskMetadata)) {
```

Comment:
I feel like we can't reuse the function here because `bucketEndTime` is exclusive while `preSelectedSegment.getEndTimeMs()` is inclusive.

Reply:
fixed